google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/ConnectionImplTest.java

```diff
@@ -412,6 +412,140 @@ public void testExecuteSetAutocommitOff() {
     }
   }
 
+  @Test
+  public void testSetAutocommitToTrue_inAutoCommitAndNotInTransaction_noop() {
+    try (ConnectionImpl subject =
+        createConnection(
+            ConnectionOptions.newBuilder()
+                .setCredentials(NoCredentials.getInstance())
+                .setUri(URI)
+                .build())) {
+      assertThat(subject.isAutocommit(), is(true));
+
+      subject.setAutocommit(true);
+
+      assertThat(subject.isAutocommit(), is(true));
+    }
+  }
+
+  @Test
+  public void testSetAutocommitToTrue_inAutoCommitAndInTransaction_noop() {
+    try (ConnectionImpl subject =
+        createConnection(
+            ConnectionOptions.newBuilder()
+                .setCredentials(NoCredentials.getInstance())
+                .setUri(URI)
+                .build())) {
+      assertThat(subject.isAutocommit(), is(true));
+      subject.execute(Statement.of("begin transaction"));
+
+      subject.setAutocommit(true);
+
+      assertThat(subject.isAutocommit(), is(true));
+    }
+  }
+
+  @Test
+  public void testSetAutocommitToFalse_inAutoCommitAndNotInTransaction_autocommitModeChanged() {
+    try (ConnectionImpl subject =
+        createConnection(
+            ConnectionOptions.newBuilder()
+                .setCredentials(NoCredentials.getInstance())
+                .setUri(URI)
+                .build())) {
+      assertThat(subject.isAutocommit(), is(true));
+
+      subject.setAutocommit(false);
+
+      assertThat(subject.isAutocommit(), is(false));
+    }
+  }
+
+  @Test
+  public void testSetAutocommitToFalse_inAutoCommitAndInTransaction_throwsException() {
+    try (ConnectionImpl subject =
+        createConnection(
+            ConnectionOptions.newBuilder()
+                .setCredentials(NoCredentials.getInstance())
+                .setUri(URI)
+                .build())) {
+      assertThat(subject.isAutocommit(), is(true));
+      subject.execute(Statement.of("begin transaction"));
+
+      subject.setAutocommit(false);
+      fail("Cannot set autocommit while in a temporary transaction");
```

Comment:
nit: use `assertThrows(..)` for this.

I know that there are examples in this test class that use this setup with a try-fail block. Those are however left-overs from when this library had to support Java 7, and we did not have support for lambda expressions.

Reply:
Done